File: backend/app/schemas/recipes.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, computed_field

from ..services.images import resolve_image_url
from ..utils.format_qty import format_qty as _format_qty


def pick_display_name(name_en: str, name_fr: str | None, language: str) -> str:
    return name_fr if (language == "fr" and name_fr) else name_en


# ── Shared mixins ──────────────────────────────────────────────────────────────

class NutritionFields(BaseModel):
    kcal: int | None = Field(default=None, ge=0)
    protein_g: int | None = Field(default=None, ge=0)
    carbs_g: int | None = Field(default=None, ge=0)
    fat_g: int | None = Field(default=None, ge=0)


class BilingualNamedItem(BaseModel):
    name: str
    name_fr: str | None = None
    display_name: str | None = None


class RecipeBase(NutritionFields):
    id: int
    title: str
    url: str
    source_domain: str
    image_local_path: str | None
    image_original_url: str | None = None
    base_servings: int
    prep_time_minutes: int | None
    is_favorite_by_me: bool = False
    recommenders: list[RecommenderOut] = []
    tags: list["TagOut"] = []  # noqa: F821 — resolved at model_rebuild time

    model_config = ConfigDict(from_attributes=True)


# ── Creation / patch ─────────────────────────────────────────────────────────

class RecipeIngredientCreate(BaseModel):
    raw_string: str = Field(min_length=1, max_length=400)
    quantity: float = Field(default=1, ge=0)
    unit: str = Field(default="item", max_length=30)
    ingredient_name: str = Field(min_length=1, max_length=200)
    ingredient_name_fr: str | None = Field(default=None, max_length=200)
    category: str = Field(default="Other", max_length=80)


class RecipeCreate(NutritionFields):
    title: str = Field(min_length=1, max_length=300)
    instructions_text: str = Field(default="", max_length=50_000)
    base_servings: int = Field(default=2, ge=1, le=100)
    prep_time_minutes: int | None = Field(default=None, ge=0, le=1440)
    ingredients: list[RecipeIngredientCreate] = Field(default_factory=list, max_length=100)


class RecipePatch(NutritionFields):
    instructions_text: str | None = Field(default=None, max_length=50_000)
    prep_time_minutes: int | None = Field(default=None, ge=0, le=1440)


# ── Output ────────────────────────────────────────────────────────────────────

class RecipeIngredientOut(BaseModel):
    raw_string: str
    quantity: float
    unit: str
    needs_review: bool
    ingredient_name: str | None = None
    ingredient_name_fr: str | None = None
    display_name: str | None = None
    category: str | None = None

    @computed_field
    @property
    def quantity_display(self) -> str:
        return _format_qty(self.quantity)


class InstructionStep(BaseModel):
    text: str
    image_url: str | None = None


class RecommenderOut(BaseModel):
    username: str
    avatar_url: str | None = None


class RecipeListItem(RecipeBase):
    @computed_field
    @property
    def image_url(self) -> str | None:
        return resolve_image_url(self.image_local_path, self.image_original_url, thumb=True)


class RecipeDetail(RecipeBase):
    instructions_text: str
    ingredients: list[RecipeIngredientOut]
    instruction_steps: list[InstructionStep] = []

    @computed_field
    @property
    def image_url(self) -> str | None:
        return resolve_image_url(self.image_local_path, self.image_original_url, thumb=False)


class RecipeTagsUpdate(BaseModel):
    tag_ids: list[int]


# Avoid circular import: TagOut is defined in tags.py and injected here.
# The annotation string "TagOut" is fine because Pydantic resolves it lazily.
from .tags import TagOut  # noqa: E402 — needed for model_rebuild

RecipeListItem.model_rebuild()
RecipeDetail.model_rebuild()

File: frontend/src/types/recipes.ts
import type { Tag } from "./tags";

export type Recommender = {
  username: string;
  avatar_url: string | null;
};

export type HFSearchResult = {
  id: string;
  name: string;
  image_url: string | null;
  tags: string[];
  total_time_minutes: number | null;
  hf_url: string;
  already_imported: boolean;
  kcal: number | null;
  protein_g: number | null;
  carbs_g: number | null;
  fat_g: number | null;
};

export type RecipeListItem = {
  id: number;
  title: string;
  url: string;
  source_domain: string;
  image_local_path: string | null;
  image_original_url: string | null;
  image_url: string | null;
  base_servings: number;
  prep_time_minutes: number | null;
  kcal: number | null;
  protein_g: number | null;
  carbs_g: number | null;
  fat_g: number | null;
  is_favorite_by_me: boolean;
  recommenders: Recommender[];
  tags: Tag[];
};

export type RecipeIngredientOut = {
  raw_string: string | null;
  quantity: number | null;
  unit: string | null;
  needs_review: boolean;
  ingredient_name: string | null;
  ingredient_name_fr: string | null;
  display_name: string | null;
  quantity_display: string;
  category: string | null;
};

export type InstructionStep = {
  text: string;
  image_url?: string | null;
};

export type RecipeDetail = RecipeListItem & {
  instructions_text: string | null;
  prep_time_minutes: number | null;
  ingredients: RecipeIngredientOut[];
  instruction_steps: InstructionStep[];
};

export type RecipeIngredientCreate = {
  raw_string: string;
  quantity: number;
  unit: string;
  ingredient_name: string;
  ingredient_name_fr?: string | null;
  category?: string;
};

export type RecipeCreate = {
  title: string;
  instructions_text?: string;
  base_servings?: number;
  prep_time_minutes?: number | null;
  ingredients?: RecipeIngredientCreate[];
};
